### PR TITLE
Implement custom rate limiter

### PR DESCRIPTION
**Pull request by FedericoCeratto**
Supports daily, weekly and monthly quotas. Quotas are measured in API usage time.
Provide stubs to add token-based quotas.

Related to https://github.com/ooni/backend/issues/371

### CODE REVIEW

File: rate_limit_quotas.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+    def consume_quota(self, elapsed: float, ipaddr=None, token=None) -> None:
+        """Consume quota in seconds
+        """
+        if ipaddr:
+            for n, limit in enumerate(self._ipaddr_limits):
+                b = self._ipaddr_buckets[n]
+                b[ipaddr] = b.get(ipaddr, limit) - elapsed
+
+        print([b[ipaddr] for b in self._ipaddr_buckets])
```
Comment (hellais) [inline rate_limit_quotas.py]: I suppose this is left over from debugging. Maybe we can either remove it or use the proper logging module for it?

File: rate_limit_quotas.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+from flask import request, current_app
+import flask
+
+
+class FlaskLimiter:
+    def _get_client_ipaddr(self):
+        # https://github.com/alisaifee/flask-limiter/issues/41
+        for m in self._limiter._ipaddr_extraction_methods:
+            if m == "X-Forwarded-For":
```
Comment (hellais) [inline rate_limit_quotas.py]: In `ipaddr_methods` it's mentioned only `["X-Real-Ip", "socket"]`, but `X-Forwarded-For` is not present, is that intended?

File: rate_limit_quotas.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        if ipaddr:
+            for n, limit in enumerate(self._ipaddr_limits):
+                b = self._ipaddr_buckets[n]
+                b[ipaddr] = b.get(ipaddr, limit) - elapsed
+
+        print([b[ipaddr] for b in self._ipaddr_buckets])
+        self.increment_quota_counters(1)
+
+    def get_minimum_across_quotas(self, ipaddr=None, token=None) -> int:
+        if ipaddr:
```
Comment (hellais) [inline rate_limit_quotas.py]: Should we maybe handle the case in which `ipaddr` is not set? Is there any way this function could get called without a valid `ipaddr` value?

File: rate_limit_quotas.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    def _after_request_callback(self, response):
+        """Consume quota and injects HTTP headers when responding to a request
+        """
+        log = current_app.logger
+        try:
+            assert response
+            tdelta = time.monotonic() - self._request_start_time
+            ipaddr = self._get_client_ipaddr()
+            self._limiter.consume_quota(tdelta, ipaddr=ipaddr)
+            q = self._limiter.get_minimum_across_quotas(ipaddr=ipaddr)
```
Comment (hellais) [inline rate_limit_quotas.py]: Maybe the missing ipaddr value can be handled here too.

File: rate_limit_quotas.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+                v += vdelta
+                if v >= limit:
+                    to_delete.append(k)
+                else:
+                    bucket[k] = v
+
+            for k in to_delete:
+                del bucket[k]
+
+    def refresh_quota_counters_if_needed(self):
```
Comment (hellais) [inline rate_limit_quotas.py]: I see that this is only called in the constructor, shouldn't it be called also periodically in some way?

File: rate_limit_quotas.py (hunk 1 above)
Comment (FedericoCeratto) [reply rate_limit_quotas.py:88]: Yes, fixed.

File: rate_limit_quotas.py (hunk 2 above)
Comment (FedericoCeratto) [reply rate_limit_quotas.py:118]: Yes: we are currently using X-Real-Ip set by Nginx

File: rate_limit_quotas.py (hunk 3 above)
Comment (FedericoCeratto) [reply rate_limit_quotas.py:92]: Not a risk with the current codebase, but I'm adding assertions around this.

File: rate_limit_quotas.py (hunk 5 above)
Comment (FedericoCeratto) [reply rate_limit_quotas.py:72]: Yes, this will be added externally to the module.

**Comment by hellais** [comment]
I see this branch is still in draft state. I also don't see any new commits since my comments, @FedericoCeratto did you maybe forget to push?

**Pull request merged**